jastrow.py

```diff
@@ -201,63 +258,100 @@ def updateinternals(self,e,epos,mask=None):
         #update b and c sums. This overlaps with testvalue()
         if mask is None:
             mask=[True]*self._configscurrent.shape[0]
+        self._bvalues[mask,:,:]+=self._get_deltab(e,epos)[mask,:,:]
+        self._avalues[mask,:,:,:]+=self._get_deltaa(e,epos)[mask,:,:,:]
         self._configscurrent[mask,e,:]=epos[mask,:]
-        self._bvalues[mask,:]+=self._get_deltab(e,epos)[mask,:]
 
 
     def value(self): 
         """  """
-        u=np.einsum("ij,j->i",self._bvalues,self.parameters['bcoeff'])+\
-          np.sum(self._avalues*self.parameters['acoeff'], axis=(2,1))
+        u=np.sum(self._bvalues*self.parameters['bcoeff'], axis=(2,1)) +\
+          np.sum(self._avalues*self.parameters['acoeff'], axis=(3,2,1))
         return (1,u)       
 
 
     def gradient(self,e,epos):
-        """We compute the gradient for electron e as 
+        """We compute the gradient for electron e as
         :math:`\grad_e \ln \Psi_J = \sum_k c_k \sum_{j > e} \grad_e b_k(r_{ej})  + \sum_{i < e} \grad_e b_k(r_{ie}) `
-        So we need to compute the gradient of the b's for these indices. 
+        So we need to compute the gradient of the b's for these indices.
         Note that we need to compute distances between electron position given and the current electron distances.
         We will need this for laplacian() as well"""
         nconf=epos.shape[0]
         ne=self._configscurrent.shape[1]
+        nup = self._mol.nelec[0]
         dnew=eedist_i(self._configscurrent,epos)
 
+        dinew=eidist_i(self._mol.atom_coords(),epos)
+        dinew=dinew.reshape(-1,3)
+
         mask=[True]*ne
         mask[e]=False
         dnew=dnew[:,mask,:]
-        dnew=dnew.reshape(-1,3)
-        
+
         delta=np.zeros((3,nconf))
+        if(e < nup): # Spin up electron selected
```

Comment:
Instead of having 'if statements littered in your code, compute the decision once and compute whatever quantity you need to change. If you need to change several things based on a conditional, do it all at once. like
if e < nup:
   nfirst=X
  cindee=0
  cindei=0
else:

Reply:
Precomputed the boolean in the if statement and used that as index